lib/spack/spack/solver/concretize.lp

```diff
@@ -461,13 +461,58 @@ error(100, "Attempted to use external for '{0}' which does not satisfy any confi
 % Config required semantics
 %-----------------------------------------------------------------------------
 
-activate_requirement_rules(Package) :- attr("node", Package).
-activate_requirement_rules(Package) :- attr("virtual_node", Package).
+activate_requirement(Package, X) :-
+  attr("node", Package),
+  requirement_group(Package, X),
+  not requirement_conditional(Package, X).
+
+activate_requirement(Package, X) :-
+  attr("virtual_node", Package),
+  requirement_group(Package, X),
+  not requirement_conditional(Package, X).
+
+activate_requirement(Package, X) :-
+  attr("node", Package),
+  1 { condition_holds(Y) : requirement_group_member(Y, Package, X) },
+  requirement_group(Package, X),
+  requirement_conditional(Package, X).
+
+activate_requirement(Package, X) :-
+  attr("virtual_node", Package),
+  1 { condition_holds(Y) : requirement_group_member(Y, Package, X) },
+  requirement_group(Package, X),
+  requirement_conditional(Package, X).
+
+{ activate_requirement_rule(Y, Package, X) : requirement_group_member(Y, Package, X) } :-
+  activate_requirement(Package, X).
+
+impose(Y) :- activate_requirement_rule(Y, Package, X).
+do_not_impose(Y) :-
+  not activate_requirement_rule(Y, Package, X),
+  requirement_group_member(Y, Package, X).
```

Comment:
From experimenting this seems to be exclusively used to prompt the correct error message: I'm worried about complicating the `.lp` logic for that purpose without a holistic approach to handing error messages (the tests make sure we get this message, but is that truly the best message to get every time there is a failure and a requirement group is involved?).

Reply:
> From experimenting this seems to be exclusively used to prompt the correct error message:

No, it's supposed to be part of the logic. A conditional requirement has the form (in the most general case with `one_of` or `any_of`):
```yaml
packages:
  zlib:
    require:
    - one_of: ["%gcc", "%clang", "%intel"]
      when: "@1.2.13"
```
The condition is the same for all the requirements, so if I look at the ASP facts for the yaml above I'll get:
```prolog
requirement_group("zlib",0).
requirement_policy("zlib",0,"one_of").
requirement_conditional("zlib",0).
condition(5,"None").
condition_requirement(5,"node","zlib").
condition_requirement(5,"node_version_satisfies","zlib","1.2.13").
imposed_constraint(5,"node_compiler_set","zlib","gcc").
imposed_constraint(5,"node_compiler_version_satisfies","zlib","gcc",":").
requirement_group_member(5,"zlib",0).
requirement_has_weight(5,0).
condition(6,"None").
condition_requirement(6,"node","zlib").
condition_requirement(6,"node_version_satisfies","zlib","1.2.13").
imposed_constraint(6,"node_compiler_set","zlib","clang").
imposed_constraint(6,"node_compiler_version_satisfies","zlib","clang",":").
requirement_group_member(6,"zlib",0).
requirement_has_weight(6,1).
condition(7,"None").
condition_requirement(7,"node","zlib").
condition_requirement(7,"node_version_satisfies","zlib","1.2.13").
imposed_constraint(7,"node_compiler_set","zlib","intel").
imposed_constraint(7,"node_compiler_version_satisfies","zlib","intel",":").
requirement_group_member(7,"zlib",0).
requirement_has_weight(7,2).
```
The `condition_requirement` is the same for each of the `one_of` requirements, so if one of them holds all of them holds[^1]. The logic here allows us to multiplex what we want to impose, and deactivate the rest.

[^1]: The `condition_requirement` maps to the single `when` in the object and defaults to "the node is present" for non-conditional requirements.

Reply:
I'll check how hard is to do that, but note that I am using the same mechanism that externals use to trim dependencies - so something we already do on ASP conditions.

Reply:
Reworked using a single condition in edd5d89b3ba4028e5232a8f3a5bb525c06b53297 The facts for the case above now look like:
```prolog
requirement_group("zlib",0).
requirement_policy("zlib",0,"one_of").
condition(5,"condition to activate requirement 0").
condition_requirement(5,"node","zlib").
condition_requirement(5,"node_version_satisfies","zlib","0.2.13").
requirement_conditional("zlib",0,5).

condition(6,"None").
condition_requirement(6,"node","zlib").
condition_requirement(6,"node_compiler","zlib","gcc").
condition_requirement(6,"node_compiler_version_satisfies","zlib","gcc",":").
imposed_constraint(6,"node_compiler_set","zlib","gcc").
imposed_constraint(6,"node_compiler_version_satisfies","zlib","gcc",":").
requirement_group_member(6,"zlib",0).
requirement_has_weight(6,0).

condition(7,"None").
condition_requirement(7,"node","zlib").
condition_requirement(7,"node_compiler","zlib","clang").
condition_requirement(7,"node_compiler_version_satisfies","zlib","clang",":").
imposed_constraint(7,"node_compiler_set","zlib","clang").
imposed_constraint(7,"node_compiler_version_satisfies","zlib","clang",":").
requirement_group_member(7,"zlib",0).
requirement_has_weight(7,1).

condition(8,"None").
condition_requirement(8,"node","zlib").
condition_requirement(8,"node_compiler","zlib","intel").
condition_requirement(8,"node_compiler_version_satisfies","zlib","intel",":").
imposed_constraint(8,"node_compiler_set","zlib","intel").
imposed_constraint(8,"node_compiler_version_satisfies","zlib","intel",":").
requirement_group_member(8,"zlib",0).
requirement_has_weight(8,2).
```